### PR TITLE
26812: Added logging to various functions and endpoints

### PR DESCRIPTION
**Pull request by cmoesgaard**

### CODE REVIEW

File: flask_saml_sso/sso.py [hunk 1]
```diff
@@ -124,23 +134,36 @@ def api_token():
 
     If no user is currently logged in, redirect to SSO flow and return here
     """
+    logger.info('API-token called')
```
Comment (danchr) [inline flask_saml_sso/sso.py]: I think this is a bit too much for the `INFO` level, but perhaps useful for `DEBUG`. Wouldn't it be better if `INFO` merely logged the _result_ of each invocation?

File: flask_saml_sso/sso.py [hunk 2]
```diff
@@ -179,9 +205,16 @@ def sso(auth):
 
     Redirects user to IdP login page specified in metadata
     """
+    logger.info('SSO called')
+
     return_to = flask.request.args.get(
         'next', flask.request.host_url)
     login = auth.login(return_to=return_to)
+
+    logger.debug('SSO - RelayState: {}'.format(return_to))
```
Comment (danchr) [inline flask_saml_sso/sso.py]: FWIW the logging module has some [built-in support](https://docs.python.org/3/library/logging.html#logging.debug) for formatting.

File: flask_saml_sso/sso.py (hunk 1 above)
Comment (cmoesgaard) [reply flask_saml_sso/sso.py]: What do you mean by result?

File: flask_saml_sso/sso.py [hunk 3]
```diff
@@ -74,8 +77,13 @@ def _get_saml_settings(app):
     s.setdefault('idp', {}).update(remote.get('idp'))
 
     if requests_signed:
+        if not os.path.isfile(cert_file):
+            logger.error('Unable to read cert file {}'.format(cert_file))
```
Comment (danchr) [inline flask_saml_sso/sso.py]: Erm, shouldn't we handle this error case somehow? Or perhaps handle the exception instead?

Comment (cmoesgaard) [reply flask_saml_sso/sso.py]: We should. I'll handle the exception.

File: flask_saml_sso/sso.py (hunk 2 above)
Comment (cmoesgaard) [reply flask_saml_sso/sso.py]: From what I can tell, the formatting is set on a global level similar to `level`, and not on a per-logger basis. I've changed the messages to be slightly less verbose, though.

**Comment by cmoesgaard** [comment]
As far as I can tell, people do both, i.e. use the app logger and a separate logger. The drawback to using the app logger is that it hides which module is performing the logging, as also makes it a bit more difficult to control the logging output _only_ for the SSO module, if we so choose.

File: flask_saml_sso/sso.py (hunk 1 above)
Comment (danchr) [reply flask_saml_sso/sso.py]: Ideally, we'd `INFO` one line for each request, and that line would tell you everything™ you need to know. `X called` seems like `DEBUG` to me.

**Comment by danchr** [comment]
> As far as I can tell, people do both, i.e. use the app logger and a separate logger. The drawback to using the app logger is that it hides which module is performing the logging, as also makes it a bit more difficult to control the logging output _only_ for the SSO module, if we so choose.

You're right, I found [the documentation](https://docs.python.org/3/library/logging.html#logging.Logger.getChild) after asking the question:

> Depending on the situation, an extension may choose to log to [`app.logger`](
http://flask.pocoo.org/docs/dev/api/#flask.Flask.logger) or its own named logger. Consult each extension’s documentation for details.

Well, whatever we do, let's document it 🙂

File: README.rst [hunk 4]
```diff
@@ -114,3 +114,13 @@ Config key                          Default             Description
...
+
+Logging
+-------
+
+The module performs extensive logging using the Flask app logging mechanism,
+through a child logger named ``sso``.
+
+- ``INFO`` provides a rough overview of the control flow.
+- ``DEBUG`` additionally provides an overview of relevant variables and the
+various request and response XML payloads.
```
Comment (danchr) [inline README.rst]: How about errors and warnings? Shouldn't we mention those as well?

File: flask_saml_sso/sso.py [hunk 5]
```diff
@@ -74,16 +80,27 @@ def _get_saml_settings(app):
...
     if requests_signed:
-        with open(cert_file, 'r') as cf:
-            cert = cf.read()
-        with open(key_file, 'r') as kf:
-            key = kf.read()
+        try:
+            with open(cert_file, 'r') as cf:
+                cert = cf.read()
+        except OSError:
+            logger.error('Unable to read cert file {}'.format(cert_file))
```
Comment (danchr) [inline flask_saml_sso/sso.py]: Include exception info as well?
```suggestion
            logger.exception('Unable to read cert file {}'.format(cert_file))
```

File: flask_saml_sso/sso.py [hunk 6]
```diff
@@ -74,16 +80,27 @@ def _get_saml_settings(app):
...
+            raise
+
+        try:
+            with open(key_file, 'r') as kf:
+                key = kf.read()
+        except OSError:
+            logger.error('Unable to read key file {}'.format(key_file))
```
Comment (danchr) [inline flask_saml_sso/sso.py]: Ditto.
```suggestion
            logger.exception('Unable to read key file {}'.format(key_file))
```

File: flask_saml_sso/sso.py [hunk 7]
```diff
@@ -224,15 +283,27 @@ def slo(auth):
 
     Redirects user to IdP SLO specified in metadata
     """
+    logger = _get_logger()
+    logger.info('SLO called')
```
Comment (danchr) [inline flask_saml_sso/sso.py]:
```suggestion
    logger.debug('SLO called')
```

File: tests/test_sso.py [hunk 8]
```diff
@@ -16,6 +16,7 @@
 from onelogin.saml2.utils import OneLogin_Saml2_Utils as saml_utils
 
 import flask_saml_sso
+from flask_saml_sso import sso
```
Comment (danchr) [inline tests/test_sso.py]: Unused?
```suggestion
```

File: flask_saml_sso/sso.py [hunk 9]
```diff
@@ -244,13 +315,24 @@ def sls(auth):
...
+        logger.error('Errors: {}'.format(errors))
         return _build_error_response(errors)
     if url is not None:
-        return flask.redirect(url)
-    return flask.redirect('/')
+        redirect_to = url
+    else:
+        redirect_to = '/'
+
+    logger.info('Redirecting to {}'.format(redirect_to))
```
Comment (danchr) [inline flask_saml_sso/sso.py]: Perhaps explain _why_ we're redirecting? E.g. “Redirecting to {} for login”?

File: flask_saml_sso/sso.py [hunk 10]
```diff
@@ -224,15 +283,27 @@ def slo(auth):
...
-        return flask.redirect(logout)
+        logger.debug('SLO Request XML: \n{}'.format(
+            auth.get_last_request_xml()))
+        redirect_to = logout
     else:
         flask.session.clear()
-        return flask.redirect('/')
+        redirect_to = '/'
+
+    logger.info('Redirecting to {}'.format(redirect_to))
```
Comment (danchr) [inline flask_saml_sso/sso.py]: Perhaps explain _why_ we're redirecting? E.g. “Redirecting to {} for logout”?

File: flask_saml_sso/sso.py [hunk 11]
```diff
@@ -244,13 +315,24 @@ def sls(auth):
...
+    logger = _get_logger()
+    logger.debug('SLS called')
+
     url = auth.process_slo(delete_session_cb=lambda: flask.session.clear())
+    logger.debug('SLS Response XML: \n{}'.format(
+        auth.get_last_response_xml()))
+
     errors = auth.get_errors()
     if errors:
+        logger.error('Errors: {}'.format(errors))
```
Comment (danchr) [inline flask_saml_sso/sso.py]: What are we doing here? A sudden “Errors:” in the log might not be terribly helpful…

Also; you're invoking `process_slo` in `sls`; is that intentional?

Comment (cmoesgaard) [reply flask_saml_sso/sso.py]: Yes.

Comment (danchr) [reply flask_saml_sso/sso.py]: > Yes.

That's vaguely suprising — a comment would be awfully nice :)

File: flask_saml_sso/sso.py (hunk 9 above)
Comment (cmoesgaard) [reply flask_saml_sso/sso.py]: Redirecting is a pretty core part of SSO/SLO as the whole flow is web based, so I thought the _why_ of it was self explanatory.

Comment (danchr) [reply flask_saml_sso/sso.py]: > Redirecting is a pretty core part of SSO/SLO as the whole flow is web based, so I thought the _why_ of it was self explanatory.

You're right, but I figured the logs targeted someone who in operations who might not be intimately familiar with it?

Comment (cmoesgaard) [reply flask_saml_sso/sso.py]: I'll add some descriptions.

File: flask_saml_sso/sso.py [hunk 12]
```diff
@@ -124,23 +140,37 @@ def api_token():
...
         attrs = flask.session.get(session.SAML_ATTRIBUTES)
+
+        logger.debug(
+            'SAML_API_TOKEN_RESTRICT_ATTR: {}'.format(group))
+        logger.debug(
+            'SAML_API_TOKEN_RESTRICT_VALUE: {}'.format(value))
+        logger.debug('SAML Attributes: {}'.format(attrs))
+
         if not attrs.get(group) or value not in attrs.get(group):
+            logger.info('User not permitted to create API tokens')
```
Comment (danchr) [inline flask_saml_sso/sso.py]: Should this be an error or warning?

File: README.rst (hunk 4 above)
Comment (cmoesgaard) [reply README.rst]: We have no warnings, and stating that `ERROR` will output any errors that occur seems a bit obvious :smile:

File: flask_saml_sso/sso.py (hunk 12 above)
Comment (cmoesgaard) [reply flask_saml_sso/sso.py]: I think errors or warnings should be used to communicate unexpected behavior in the application. This is a scenario we specifically designed it to handle.

File: README.rst (hunk 4 above)
Comment (cmoesgaard) [reply README.rst]: But sure, I'll add something for consistency I guess.

File: flask_saml_sso/sso.py [hunk 13]
```diff
@@ -124,23 +140,37 @@ def api_token():
...
+
+        logger.info('User not logged in, redirecting to {}'.format(
+            redirect_url))
+        return flask.redirect(redirect_url)
 
     app = flask.current_app
     # Check if user is permitted to create API tokens
 
     if app.config.get('SAML_API_TOKEN_RESTRICT', False):
+        logger.info('Token restriction enabled')
```
Comment (danchr) [inline flask_saml_sso/sso.py]:
```suggestion
        logger.debug('Token restriction enabled')
```

File: flask_saml_sso/sso.py (hunk 12 above)
Comment (danchr) [reply flask_saml_sso/sso.py]: > I think errors or warnings should be used to communicate unexpected behavior in the application. This is a scenario we specifically designed it to handle.

Indeed, but I think we should consider warning about it; someone may be trying to do something they shouldn't be doing.

Comment (cmoesgaard) [reply flask_saml_sso/sso.py]: That's fair. Let's make it a warning then.

Comment (cmoesgaard) [inline flask_saml_sso/sso.py]:
```suggestion
            logger.warning('User not permitted to create API tokens')
```

**Pull request merged**